# biosample-basex
_Using XQueries in a BaseX database to convert NCBI's BioSample database from XML to SQLite._

Querying the native BioSample database requires familiarity with, and tools designed for XPath, XQuery or XSLT. The SQLite database generated by this software can be queried from Python programs or a multitude of commercial, open-source, CLI and GUI tools, using _the most common query language_. No citation provided :-)

**Requires SQLite3 3.32**

## BioSample metadata background

When a sicentist submits data (like taxonomic identification or gene expression) to NCBI, they are also required to submit metadata about the [BioSamples](https://www.ncbi.nlm.nih.gov/biosample/) from which that data was collected. 

NCBI uses the term **attribute** to describe certain named information about the BioSamples. Submitters can tag this information about their BioSamples with any attribute names they choose. The NCBI performs curation of the BioSamples, and attributes that appear to share the same meaning as controlled terms from standards like the [GSC's `MIxS`](https://gensc.org/mixs/) are given harmonized names.

These BioSample `attribute`s should not be confused with XML markup construct that is also called [`attribute`](https://en.wikipedia.org/wiki/XML#Key_terminology). In fact, the BioSample attributes are actually XML `element`s (from the path `BioSampleSet/BioSample/Attributes/Attribute`) although `@attribute_name` and `@harmonized_name` are XML `attribute`s.

There are several other paths with `element`s, and `attribute`s of elements that desribe the BioSamples. See XXX.

## BaseX database limits

As of early January, 2022, the number of XML nodes necessary to model the 22,786,924 BioSamples was 2,397,333,775. This repo uses the BaseX open-source XML database, which is feature rich and performs well. However, BaseX does have per-database [limits](https://docs.basex.org/wiki/Statistics), such as no more than 2,147,483,648 nodes per database. Therefore, NCBI's [biosample_set.xml.gz](https://ftp.ncbi.nlm.nih.gov/biosample/) is split into two chunks with `sed` and then loaded in two BaseX databases. Queries written in the `XQuery` language are then executed over the two databases, generating various `TSV` files.  


```
Name             Resources  Size         Input Path                                                                                           
--------------------------------------------------------------------------------------------------------------------------------------------
biosample_set_1  1          32517967506  /global/cfs/cdirs/m3513/endurable/biosample/biosample-basex/target/biosample_set_under_12500001.xml  
biosample_set_2  1          30367728281  /global/cfs/cdirs/m3513/endurable/biosample/biosample-basex/target/biosample_set_over_12500001.xml   
```

## SQLite output

After running the XQueries, the generated `TSV`s are loaded into a SQLite database. SQLite was selected becasue the entire, indexed and ready-to-use database can be shared as a single, compressable file and beacuse of the plethora of compatible tools, which do not require setting up a persistent database server. Examples include Python's built-in [sqlite module](https://docs.python.org/3/library/sqlite3.html), the `sqlite` command line client (which can be dowloaded from https://www.sqlite.org/download.html or installed with package managers that are included in many Unix-like operating systems), or with a graphical tool like [DBeaver](https://dbeaver.io/download/). 

The 

- all_attribs
- env_package_repair
- harmonized_wide
- harmonized_wide_repaired
- non_attribute_metadata

As of 2022-01-03
_The SQLite DBs contain minimal indices_
- biosample_basex.db.gz: 7.7 GB
- biosample_basex.db:   67.9 GB

## Usage

BaseX can be installed with homebrew on Macs or `apt-get` on Ubuntu Linux machines. It's open-source software.

Downloading the `.zip` archive makes it a little more straightforward to increase the amount of memory allocated by the launch scripts. For example, on a 32 GB machine, `basex/bin/basex` might look like this:

```bash
#!/usr/bin/env bash

# Path to this script
FILE="${BASH_SOURCE[0]}"
while [ -h "$FILE" ] ; do
  SRC="$(readlink "$FILE")"
  FILE="$( cd -P "$(dirname "$FILE")" && \
           cd -P "$(dirname "$SRC")" && pwd )/$(basename "$SRC")"
done
MAIN="$( cd -P "$(dirname "$FILE")/.." && pwd )"

# Core and library classes
CP=$MAIN/BaseX.jar:$MAIN/lib/custom/*:$MAIN/lib/*:$CLASSPATH

# Options for virtual machine (can be extended by global options)
BASEX_JVM="-Xmx24g $BASEX_JVM"

# Run code
exec java -cp "$CP" $BASEX_JVM org.basex.BaseX "$@"
```

https://docs.basex.org/wiki/Main_Page

https://basex.org/download/


## NERSC cori specific notes

`make all` takes roughly 8 hours.

- `cd /global/cfs/cdirs/m3513/endurable/biosample`
- `git clone git@github.com:turbomam/biosample-basex.git` 
- `wget` the latest BaseX zip archive
    - 2022-01-02: `wget https://files.basex.org/releases/9.6.4/BaseX964.zip`
- `unzip BaseX964.zip`
- increase the Java memory allocation in `basex/bin/basex`
    - `BASEX_JVM="-Xmx96g $BASEX_JVM"`
    - That may be more than necessary. Seems to run OK on an Intel MacBook Pro with 24 GB allocated.
- `cp biosample-basex/template.env biosample-basex/.env`
- edit `biosample-basex/.env`
    - `BASEXCMD = ../basex/bin/basex`
    - The value for `del_from` should be roughly half the number of BioSamples. The suggested value of 12500001 should be reasonable but could be revised after downloading and unpacking the BioSample XML file. Run something like `tail -n 50 target` at the shell prompt, note the `id` attribute of the last `<BioSample>`, and set `del_from` to an integer close to one half of that last `id` value. 
- Start a session manager like `screen`. If you get disconnected, it may be tricking to log back inot the same cori node because of load balancing, but scripts started before the disconnection should run to completion. It doesn't hurt to force your client computer to stay awake with something like `caffeine`.
- `module load python`
    - On cori, this makes common Python paackages like `pandas` available. On other systems, users should create a `venv` virtual environment, enter it, and run `pip install -r requirments.txt`
- cd `biosample-basex/`
- `make all`

----

results can be found at https://portal.nersc.gov/project/m3513/biosample/
